onnxruntime/core/providers/cpu/ml/tree_ensemble_common.h

```diff
@@ -0,0 +1,706 @@
+  std::vector<int64_t> class_labels_;
+
+ public:
+  TreeEnsembleCommonClassifier(int omp_tree, int omp_N) : TreeEnsembleCommon<ITYPE, OTYPE>(omp_tree, omp_N) {}
+
+  inline int64_t get_class_count() const { return class_count_; }
+
+  void init(
+      const std::string& aggregate_function,
+      std::vector<OTYPE> base_values,
+      std::vector<int64_t> nodes_falsenodeids,
+      std::vector<int64_t> nodes_featureids,
+      std::vector<OTYPE> nodes_hitrates,
+      std::vector<int64_t> nodes_missing_value_tracks_true,
+      const std::vector<std::string>& nodes_modes,
+      std::vector<int64_t> nodes_nodeids,
+      std::vector<int64_t> nodes_treeids,
+      std::vector<int64_t> nodes_truenodeids,
+      std::vector<OTYPE> nodes_values,
+      const std::string& post_transform,
+      std::vector<int64_t> class_ids,
+      std::vector<int64_t> class_nodeids,
+      std::vector<int64_t> class_treeids,
+      std::vector<OTYPE> class_weights,
+      std::vector<std::string> classlabels_strings,
+      std::vector<int64_t> classlabels_int64s);
+
+  void compute(const Tensor* X, Tensor* Z, Tensor* label) const;
+};
+
+template <typename ITYPE, typename OTYPE>
+void TreeEnsembleCommonClassifier<ITYPE, OTYPE>::init(
+    const std::string& aggregate_function,
+    std::vector<OTYPE> base_values,
+    std::vector<int64_t> nodes_falsenodeids,
+    std::vector<int64_t> nodes_featureids,
+    std::vector<OTYPE> nodes_hitrates,
+    std::vector<int64_t> nodes_missing_value_tracks_true,
+    const std::vector<std::string>& nodes_modes,
+    std::vector<int64_t> nodes_nodeids,
+    std::vector<int64_t> nodes_treeids,
+    std::vector<int64_t> nodes_truenodeids,
+    std::vector<OTYPE> nodes_values,
+    const std::string& post_transform,
+    std::vector<int64_t> class_ids,
+    std::vector<int64_t> class_nodeids,
+    std::vector<int64_t> class_treeids,
+    std::vector<OTYPE> class_weights,
+    std::vector<std::string> classlabels_strings,
+    std::vector<int64_t> classlabels_int64s) {
+  classlabels_strings_ = classlabels_strings;
+  classlabels_int64s_ = classlabels_int64s;
+  class_count_ = classlabels_strings_.size() == 0 ? classlabels_int64s_.size() : classlabels_strings_.size();
+  if (class_count_ == 0) {
+    ORT_THROW("classlabels_int64s_ and classlabels_strings_ cannot be both empty.");
+  }
+
+  TreeEnsembleCommon<ITYPE, OTYPE>::init(
+      aggregate_function, base_values, class_count_, nodes_falsenodeids,
+      nodes_featureids, nodes_hitrates, nodes_missing_value_tracks_true,
+      nodes_modes, nodes_nodeids, nodes_treeids, nodes_truenodeids,
+      nodes_values, post_transform, class_ids, class_nodeids,
+      class_treeids, class_weights);
+
+  std::set<int64_t> weights_classes;
+  weights_are_all_positive_ = true;
+  for (size_t i = 0, end = class_ids.size(); i < end; ++i) {
+    weights_classes.insert(class_ids[i]);
+    if (class_weights[i] < 0)
+      weights_are_all_positive_ = false;
+  }
```

Comment:
Probably not significant perf-wise, but moving this out of the loop and doing std::any_of would be cheaper if you expect to find negative weights, as it would stop the check at the first negative value.

Reply:
It is inserted in a loop which has to go all over the weights. I can use any_of but I need to create a second loop then. Not sure it would be faster.